consumer.go

```diff
@@ -80,49 +128,80 @@ func (c *Consumer) isConnected() bool {
 	return c.consumerGroup != nil
 }
 
-// StartListening is a blocking call that tries to establish a connection to Kafka and then starts listening.
-func (c *Consumer) StartListening(messageHandler func(message FTMessage)) {
-	if !c.isConnected() {
-		c.connect()
-	}
+func (c *Consumer) setMonitor(monitor *consumerMonitor) {
+	c.monitorLock.Lock()
+	defer c.monitorLock.Unlock()
 
-	handler := newConsumerHandler(c.logger, messageHandler)
+	c.monitor = monitor
+}
+
+// isMonitorConnected returns whether the consumer monitor is set.
+// It is only set if a successful connection is established.
+func (c *Consumer) isMonitorConnected() bool {
+	c.monitorLock.RLock()
+	defer c.monitorLock.RUnlock()
+
+	return c.monitor != nil
+}
```

Comment:
**question:** Is it possible to lose connection and what is happening then ? I'm curious because here we aren't actually checking whether we' are connected.

Reply:
The connection is handled internally by the Sarama library and is being refreshed behind the scenes.

Requests to the Kafka broker would start failing if the connection is somehow lost however we'd notice that by the response of the `MonitorCheck()` calls. Keep in mind that if this connection is lost then it also makes sense for the consumer one to fail as well so trying to re-establish it is out of the scope for us.

I see two potential root causes of such scenarios:
- the Kafka cluster having issues
- a bug in the Sarama library

If we're dealing with the latter, then I'd say we rather switch the Kafka library with something else.